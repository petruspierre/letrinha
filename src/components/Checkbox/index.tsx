import {
  CheckboxContainer,
  HiddenCheckbox,
  Icon,
  StyledCheckbox,
} from "./styles";

export interface CheckboxProps extends HTMLInputElement {
  checked: boolean;
}

const Checkbox = ({ checked, ...props }: CheckboxProps) => (
  <CheckboxContainer>
    <HiddenCheckbox checked={checked} {...props} />
    <StyledCheckbox checked={checked}>
      <Icon viewBox="0 0 24 24">
        <polyline points="20 6 9 17 4 12" />
      </Icon>
    </StyledCheckbox>
  </CheckboxContainer>
);

export default Checkbox;
