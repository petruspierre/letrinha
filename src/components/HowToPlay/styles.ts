import styled from "styled-components";
import is from "styled-is";

interface IPreviewProps {
  exists?: boolean;
  correctPlace?: boolean;
}

export const Container = styled.article`
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 0 2rem;
  color: ${({ theme }) => theme.colors.white};

  font-family: ${({ theme }) => theme.font.primary};

  p {
    font-size: 1rem;
    margin-bottom: 0.5rem;

    span {
      font-weight: 400;
      color: ${({ theme }) => theme.colors.green};
    }

    &:first-child {
      margin-bottom: 1.5rem;
    }
  }
`;

export const PreviewWrapper = styled.div`
  display: flex;
  padding: 0.5rem 0.75rem;
  gap: 0.5rem;
  background: ${({ theme }) => theme.colors.secondary};
  margin-bottom: 2rem;
`;

export const Preview = styled.div<IPreviewProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 2rem;
  height: 2.5rem;
  width: 2.5rem;
  background: ${({ theme }) => theme.colors.primary};
  font-family: ${({ theme }) => theme.font.primary};
  font-style: italic;
  color: rgba(255, 255, 255, 0.5);

  ${is("exists")`
    background: ${({ theme }) => theme.colors.yellow};
    color: ${({ theme }) => theme.colors.white};
  `}

  ${is("correctPlace")`
    background: ${({ theme }) => theme.colors.green};
    color: ${({ theme }) => theme.colors.white};
  `}
`;
