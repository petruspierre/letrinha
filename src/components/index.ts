export { default as ThemeSelection } from "./ThemeSelection";
export { default as Keyboard } from "./Keyboard";
export { default as HowToPlay } from "./HowToPlay";
export { default as Header } from "./Header";
export { default as SimpleGame } from "./SimpleGame";
export { default as Canva } from "./Canva";
