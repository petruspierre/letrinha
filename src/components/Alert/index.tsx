import { Container } from "./styles";

interface AlertProps {
  alert: Partial<{
    title?: string;
    message: string;
  }>;
  dismissAlert: () => void;
}

const Alert = ({ alert, dismissAlert }: AlertProps) => {
  return (
    <Container>
      {alert && (
        <>
          {alert.title && <h3>{alert.title}</h3>}
          <p>{alert.message}</p>
        </>
      )}
    </Container>
  );
};

export default Alert;
