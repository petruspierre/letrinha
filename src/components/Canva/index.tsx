import React from "react";
import { IGuess } from "~/model/SimpleGame";
import { Container, Field, FieldsWrapper } from "./styles";

interface ICanvaProps {
  wordLength: number;
  attempts: number;
  selectedLetter: number;
  selectedGuess: number;
  index: number;
  onLetterClick: (letter: number) => void;
  guesses: IGuess[];
  isGameOver: boolean;
}

const Canva = ({
  wordLength,
  attempts,
  selectedGuess,
  selectedLetter,
  onLetterClick,
  index,
  guesses,
  isGameOver,
}: ICanvaProps) => {
  return (
    <Container>
      <FieldsWrapper columns={wordLength} rows={attempts}>
        {guesses.map((guess, guessIndex) => {
          return (
            <React.Fragment key={guessIndex}>
              {Array(wordLength)
                .fill(null)
                .map((_, letterIndex) => {
                  const existingLetter = guess[letterIndex];
                  const isSelectedGuess = selectedGuess === guessIndex;

                  const letterExists = existingLetter?.exists;
                  const letterCorrectPlace = existingLetter?.correctPlace;

                  return (
                    <Field
                      blur={
                        (!isSelectedGuess &&
                          !letterExists &&
                          !letterCorrectPlace) ||
                        isGameOver
                      }
                      isActive={
                        selectedLetter === letterIndex &&
                        isSelectedGuess &&
                        !isGameOver
                      }
                      exists={letterExists}
                      correctPlace={letterCorrectPlace}
                      onClick={(e: any) => [
                        onLetterClick(letterIndex),
                        e.target.blur(),
                      ]}
                      disabled={!isSelectedGuess && !isGameOver}
                      key={String(letterIndex)}
                      data-testid={`field-${index}-${guessIndex}-${letterIndex}`}
                    >
                      {existingLetter?.letter ?? ""}
                    </Field>
                  );
                })}
            </React.Fragment>
          );
        })}
      </FieldsWrapper>
    </Container>
  );
};

export default Canva;
