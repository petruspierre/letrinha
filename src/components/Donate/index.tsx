import Modal from "../Modal";
import QRCode from "react-qr-code";
import { Container, QRContainer } from "./styles";

interface IDonateProps {
  dismiss: () => void;
}

const PIX_KEY = "ajude.letrinha@petrus.dev.br";

const Donate = ({ dismiss }: IDonateProps) => {
  return (
    <Modal dismiss={dismiss} title="Nos ajude a continuar melhorando!">
      <Container>
        <p>
          Nos faça uma doação de qualquer valor para nos estimular a continuar.
        </p>

        <span>QR code do pix:</span>
        <QRContainer>
          <QRCode
            value="00020126680014br.gov.bcb.pix0128ajude.letrinha@petrus.dev.br0214Ajude Letrinha5204000053039865802BR5901a6006CIDADE62070503***6304C038"
            title="Chave pix"
          />
        </QRContainer>

        <p>Chave pix por email:</p>
        <p>
          {PIX_KEY}
          <button
            title="Copiar chave"
            aria-label="Copiar chave pix"
            onClick={() => navigator.clipboard.writeText(PIX_KEY)}
          >
            Copiar
          </button>
        </p>
      </Container>
    </Modal>
  );
};

export default Donate;
