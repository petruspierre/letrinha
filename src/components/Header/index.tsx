import { FaPalette, FaQuestionCircle } from "react-icons/fa";
import { useRouter } from "next/router";

import Logo from "~/components/Logo";
import { useTheme } from "~/styles/theme";

import { Button, Container, LogoButton } from "./styles";

interface HeaderProps {
  toggleInstructions: () => void;
  showLogo: boolean;
}

const Header = ({ toggleInstructions, showLogo = true }: HeaderProps) => {
  const { showThemeSelection } = useTheme();
  const router = useRouter();

  const goHome = () => {
    router.push("/");
  };

  return (
    <Container>
      <Button
        onClick={toggleInstructions}
        aria-label="Fazer doação"
        title="Fazer doação"
      >
        <img src="/assets/icons/donate.svg" alt="Ícone de cofre" />
      </Button>
      <Button
        onClick={toggleInstructions}
        aria-label="Como jogar?"
        title="Como jogar?"
      >
        <img src="/assets/icons/help.svg" alt="Ícone de interrogação" />
      </Button>
      {showLogo && (
        <LogoButton onClick={goHome} title="Letrinha - Voltar para início">
          <Logo />
        </LogoButton>
      )}
      <Button
        onClick={showThemeSelection}
        aria-label="Mudar tema"
        title="Mudar tema"
      >
        <img src="/assets/icons/theme.svg" alt="Ícone de paleta" />
      </Button>
      <Button
        onClick={showThemeSelection}
        aria-label="Configurações"
        title="Configurações"
      >
        <img src="/assets/icons/settings.svg" alt="Ícone de engrenagem" />
      </Button>
    </Container>
  );
};

export default Header;
