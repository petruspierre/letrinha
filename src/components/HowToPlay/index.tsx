import { Container, Overlay } from "./styles";

interface HowToPlayProps {
  dismiss: () => void;
}

const HowToPlay = ({ dismiss }: HowToPlayProps) => {
  return (
    <>
      <Overlay onClick={dismiss} />
      <Container>
        <button
          aria-label="Fechar tutorial"
          title="Fechar tutorial"
          onClick={dismiss}
        >
          X
        </button>

        <h2>Como jogar?</h2>

        <p>
          Adivinha a palavra do dia acertando as <strong>letrinhas</strong> na
          quantidade limite de tentativas (quantidade de letras + 1)
        </p>

        <p>
          Para saber se você está quente ou frio, observe como ficarão as
          letrinhas após fazer sua tentativa:
        </p>

        <h3>
          Cor <span className="yellow">amarela</span>:
        </h3>
        <p>
          Quando sua letrinha ficar amarela, significa que ela está na palavra
          do dia, porém cuidado, ela não está no local certo!
        </p>

        <h3>
          Cor <span className="green">verde</span>:
        </h3>
        <p>
          Quando sua letrinha ficar verde, significa que a letrinha existe na
          palavra e já está no lugar certo!
        </p>

        <p>
          Se sua letrinha não mudar de cor, pode descartar, pois ela não faz
          parte da palavra.
        </p>

        <small>
          Lembre-se: apenas são consideradas palavras do dicionário! Nada de
          inventar nem usar nomes próprios.
        </small>
      </Container>
    </>
  );
};

export default HowToPlay;
