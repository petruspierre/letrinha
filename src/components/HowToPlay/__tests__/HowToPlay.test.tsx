import { render, fireEvent } from "@test/components";

import HowToPlay from "..";

const defaultProps = {
  dismiss: jest.fn(),
};

describe("HowToPlay component", () => {
  it("matches snapshot", () => {
    const { container } = render(<HowToPlay {...defaultProps} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("calls dismiss callback", () => {
    const cb = jest.fn();

    const { getByTitle } = render(<HowToPlay {...defaultProps} dismiss={cb} />);

    fireEvent.click(getByTitle("Fechar janela"));

    expect(cb).toBeCalledTimes(1);
  });
});
