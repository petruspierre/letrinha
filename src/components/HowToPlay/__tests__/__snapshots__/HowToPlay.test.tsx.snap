// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowToPlay component matches snapshot 1`] = `
.c0 {
  cursor: default;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,0.5);
  z-index: 25;
}

<button
  class="c0"
/>
`;
